rfd/0213-relay-a-lightweight-proxy.md

```diff
@@ -0,0 +1,64 @@
+
+All the relays of the group will serve client connections in the same way, so clients only ever need to interact directly with the single load balancer address - this is a requirement to support Kubernetes clients without a local proxy - and will do so depending on the required protocol.
+
+For SSH, the existing proxy transport protocol will also be implemented by relays, which will run the same connection router that's used by the proxy to resolve a connection target into a host ID; however, unlike the actual proxies, relays will only ever route connections to resources that are available within the same relay group and will never open TCP connections to direct dial agents, and, to avoid giving relays any more privileges than they strictly require, agentless SSH servers and proxy recording mode are not going to be supported. For consistency with the router in the Teleport proxy, host ID resolution will happen against the full SSH inventory without filtering for nodes available in the subenvironment, and the usual rules will apply in case of an ambiguous match.
+
+For Kubernetes access through a relay we will tweak the protocol to include the destination cluster name as part of the SNI of the connection, and we will tweak the host side credentials of Kubernetes service agents to also include a wildcard DNS SAN (with a suffix to be determined) to allow unmodified Kubernetes API clients to validate the server certificate of the Kubernetes service. Connections coming from the regular control plane of the cluster will remain unmodified. The relay will check the inventory of kubernetes clusters based on the name extracted from SNI, it will pick a host serving the requested cluster, and forward the TLS connection along without terminating it; similarly to SSH, the resolution will happen against the full inventory, and if the resolved host ID is not connected to the relay group then the connection will fail.
+
+### Global connection routing and relay peering
+
+Resource heartbeats from agents connected to relays will include a list of relay IDs; that list will be used by proxies to reference the relay heartbeats (which will include the proxy IDs that they're connected to); if a relay has an open tunnel to the proxy trying to open a connection, that relay will be used, otherwise the usual proxy peering rules apply when picking a peer proxy that has a tunnel to one of the relays. The proxy peering dial request will be extended to include an optional relay ID (valid when connecting to a host ID), and the peer proxy receiving the dial request will forward the connection to the specified relay through the reverse tunnel; the dial request in the reverse tunnel will include the actual target. The target must be available as a reverse tunnel to that relay directly.
```

Comment:
An alternative that was initially considered is to just have the relay group name in the heartbeat but we would then have to implement some long-running gossip mechanism for relays to know which tunnels are available in which relay, and we have nothing of the sorts in Teleport right now.